File: data structures and algorithms/cpp/08_071.cpp
#include <stack>
#include <vector>
#include <string>
#include <algorithm>
using namespace std;

// Structure to represent a clothing item
struct ClothingItem {
   string type;
   int popularity;
   
   // Constructor for clothing item
   ClothingItem(string t, int p) : type(t), popularity(p) {}
};

// Class to manage a stack of clothing items
class FashionStack {
private:
   stack<ClothingItem> mainStack;
   stack<ClothingItem> tempStack;
   
public:
   // Add a new clothing item to the stack
   void addItem(string type, int popularity) {
      mainStack.push(ClothingItem(type, popularity));
   }
   
   // Remove and return the top item from the stack
   ClothingItem removeTopItem() {
      if (mainStack.empty()) {
         return ClothingItem("", -1); // Invalid item
      }
      
      ClothingItem top = mainStack.top();
      mainStack.pop();
      return top;
   }
   
   // Find and remove most popular item of target type
   ClothingItem findAndRemoveType(string targetType) {
      if (mainStack.empty()) {
         return ClothingItem("", -1); // Invalid item
      }
      
      // Convert stack to vector for easier searching
      vector<ClothingItem> items;
      while (!mainStack.empty()) {
         items.push_back(mainStack.top());
         mainStack.pop();
      }
      
      // Find most popular item of target type
      ClothingItem* mostPopular = nullptr;
      int mostPopularIndex = -1;
      
      for (int i = 0; i < items.size(); i++) {
         if (items[i].type == targetType) {
            if (mostPopular == nullptr || 
                items[i].popularity > mostPopular->popularity) {
               mostPopular = &items[i];
               mostPopularIndex = i;
            }
         }
      }
      
      if (mostPopularIndex == -1) {
         // Restore stack and return invalid item
         for (int i = items.size() - 1; i >= 0; i--) {
            mainStack.push(items[i]);
         }
         return ClothingItem("", -1);
      }
      
      // Save the result
      ClothingItem result = items[mostPopularIndex];
      
      // Rebuild stack without the found item
      for (int i = items.size() - 1; i >= 0; i--) {
         if (i != mostPopularIndex) {
            mainStack.push(items[i]);
         }
      }
      
      return result;
   }
   
   // Reorganize stack by popularity (highest first)
   void reorganizeByPopularity() {
      if (mainStack.empty()) {
         return;
      }
      
      // Convert stack to vector
      vector<ClothingItem> items;
      while (!mainStack.empty()) {
         items.push_back(mainStack.top());
         mainStack.pop();
      }
      
      // Sort by popularity (highest first)
      sort(items.begin(), items.end(), 
           [](const ClothingItem& a, const ClothingItem& b) {
         return a.popularity > b.popularity;
      });
      
      // Rebuild stack with sorted items
      for (const auto& item : items) {
         mainStack.push(item);
      }
   }
};

File: data structures and algorithms/cs/08_071.cs
using System;
using System.Collections.Generic;

class ClothingItem
{
   public string Type { get; }
   public int Popularity { get; }

   public ClothingItem(string type, int popularity)
   {
      Type = type;
      Popularity = popularity;
   }
}

public class FashionStack
{
   private Stack<ClothingItem> mainStack;
   private Stack<ClothingItem> tempStack;

   public FashionStack()
   {
      mainStack = new Stack<ClothingItem>();
      tempStack = new Stack<ClothingItem>();
   }

   public void AddItem(string type, int popularity)
   {
      mainStack.Push(new ClothingItem(type, popularity));
   }

   public ClothingItem RemoveTopItem()
   {
      if (mainStack.Count == 0)
      {
         return null;
      }
      return mainStack.Pop();
   }

   public ClothingItem FindAndRemoveType(string targetType)
   {
      if (mainStack.Count == 0)
      {
         return null;
      }

      ClothingItem mostPopular = null;
      int mostPopularIndex = -1;

      for (int i = mainStack.Count - 1; i >= 0; i--)
      {
         ClothingItem item = mainStack.ToArray()[i];
         if (item.Type == targetType)
         {
            if (mostPopular == null ||
               item.Popularity > mostPopular.Popularity)
            {
               mostPopular = item;
               mostPopularIndex = i;
            }
         }
      }

      if (mostPopularIndex == -1)
      {
         return null;
      }

      int itemsToMove =
         mainStack.Count - mostPopularIndex - 1;
      for (int i = 0; i < itemsToMove; i++)
      {
         tempStack.Push(mainStack.Pop());
      }

      ClothingItem result = mainStack.Pop();

      while (tempStack.Count > 0)
      {
         mainStack.Push(tempStack.Pop());
      }

      return result;
   }

   public void ReorganizeByPopularity()
   {
      if (mainStack.Count == 0)
      {
         return;
      }

      while (mainStack.Count > 0)
      {
         tempStack.Push(mainStack.Pop());
      }

      List<ClothingItem> sortedItems =
         new List<ClothingItem>(tempStack);
      sortedItems.Sort((a, b) =>
         b.Popularity.CompareTo(a.Popularity));
      tempStack.Clear();

      foreach (ClothingItem item in sortedItems)
      {
         mainStack.Push(item);
      }
   }
}
